docs/for-maintainers.md

```diff
@@ -0,0 +1,423 @@
+
+Feel free to add console messages that demonstrate callback behavior, especially their function signatures.
+
+Example:
+
+```js
+const onClick = (...args) => console.log('onClick args:', ...args)
+
+export const Button = args => <Button {...args} />
+Button.args = { onClick: onClick }
+```
+
+Simply passing `console.log` to callbacks like `onClick` is probably enough.
+
+The Actions addon adds some slightly different tools for demonstrating callbacks, but the added utility is not really worth using a new system for callback logging.
+
+### Customising source code snippets
+
+Make sure the source code snippets for demos gives the user useful information. There are a few ways to show source code snippets in the Docs Page:
+
+1. 'Dynamic', JSX-like rendering of the story. This is the default if the story uses `args`, and updates in response to changes in the Args Table.
+2. A literal, nearly character-for-character expression of the story. This is the fallback if the story does not use `args`, and is the same as the code snippet shown in the 'Story' addon in the Canvas tab. This option does not dynamically update with the Args Table, but _does_ show accessory code like hooks that is useful to explain how a component requires a ref or state.
+3. A hand-written snippet. It will receive syntax highlighting when shown on the Docs Page.
+4. It's also possible to customize the source-code processing function. See [this example](https://github.com/storybookjs/storybook/blob/master/addons/docs/docs/recipes.md#customizing-source-snippets).
+
+To use #2 for a story that uses `args`, set `parameters.docs.source.type` to `'code'`:
+
+```js
+export const StatefulStory = () => {...}
+StatefulStory.parameters = { docs: { source: { type: 'code' } } }
+```
+
+This can be done at the story or component level.
+
+If neither of the auto-generated code snippets provide meaningful examples, use a handwritten snippet by setting `parameters.docs.source.code` to the string that you would like to appear in the code snippet:
+
+```js
+export const CustomCodeSnippet = () => {}
+CustomCodeSnippet.parameters = {
+    docs: {
+        source: {
+            code: `
+const ComponentDemo = () => {
+    const ref = useRef()
+    
+    return <ExampleComponent ref={ref} />
+}
+`,
+        },
+    },
+}
+```
+
+Read more about source code blocks at [Storybook: Doc Blocks - Source](https://storybook.js.org/docs/react/writing-docs/doc-blocks#source)
+
+### Making a reusable story template
+
+To keep stories concise and DRY, you can reuse a story template (that uses the `args` syntax) and set the relevant props on each story using `StoryName.args = { ... }`. That looks like this:
+
+```js
+// The story template
+const Template = args => <Button {...args} />
+
+// Stories to render
+export const Primary = Template.bind({}) // Makes a copy of the template
```

Comment:
Since arrow functions lexically bind `this`, `someArrowFn.bind({})` is equivalent to `someArrowFn.bind()` and so we don't need to pass an object as `this` when binding a template.

(I should have brought this up earlier when we were porting stories to the new storybook format :sweat_smile:)

Reply:
Ah good to know 😁 I just copied the storybook docs 😅 

After thinking about it a bit, since they're equivalent, I think there's an argument to be made for keeping `.bind({})` how it is so it's the same as the storybook docs (which are linked to a lot in this 'For Maintainers' page) so people don't wonder about the difference as they switch between SB's docs and our docs - does that sound okay to you? 😁  Maybe once everyone adjusts to the new story format we can remove the `{}` library-wide 🙂

Reply:
I've been thinking about why the Storybook team may have used `bind` in this way, and I think that it might be due to typescript: its [type definition for `bind`](https://github.com/microsoft/TypeScript/blob/16031bc429305e5daabf263f208678a6729a161e/lib/lib.es5.d.ts#L293) specifies one required argument ([code example with error](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcAUBKAXIgylAJxjAHNEAfRcAEwFNhjbrEBvAKEU8QNqhAKQAiFHEFsAvmzYQEAZyiIAHogC8yVADoARsWqYgA)). Since [there's prior discussion about us moving to typescript for more apps](https://github.com/dhis2/notes/issues/33) it makes sense to keep `.bind({})` even once everyone adjusts to CSF.

So yeah let's stick to how things are now to not only be consistent with the storybook docs and avoid confusion but also because it makes a potential switch to typescript less painful :+1:

Reply:
Ah interesting, it's nice that this supports typescript :) thanks for taking a look!